salt/minion.py

```diff
@@ -1443,6 +1443,11 @@ def _fire_master(self, data=None, tag=None, events=None, pretag=None, timeout=60
         else:
             return
 
+        if include_startup_grains:
+            grains_to_add = dict(
+                [(k, v) for k, v in six.iteritems(self.opts.get('grains', {})) if k in self.opts['start_event_grains']])
+            load['grains'] = grains_to_add
```

Comment:
Much faster, more readable version that doesn't iterates every time on each grain, doesn't require new parameter and doesn't crashes if `start_event_grains` not there, but logs configuration typos:
```python
if "grains" in self.opts and bool(self.opts.get("start_event_grains")):
    load["grains"] = {}
    for k in self.opts["start_event_grains"]:
        try:
            load["grains"][k] = self.opts["grains"][k]
        except KeyError as exc:
            log.error("Cannot add event grain: {} not found".format(k))
```